## feature engineering

In [2]:
import sys
import os

sys.path.append(os.path.abspath(".."))

from src.data_loader import load_data
from src.preprocessing import create_features

df = load_data()

df = create_features(df)

df.head()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal,RoomsPerHousehold,BedroomsRatio,PopulationPerHousehold
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526,2.732919,0.146591,126.0
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585,2.956685,0.155797,1138.0
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521,2.957661,0.129516,177.0
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413,2.283154,0.184458,219.0
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422,2.879646,0.172096,259.0


# evaluate engineered features

In [3]:
corr = df.corr(numeric_only=True)

corr["MedHouseVal"].sort_values(
    ascending=False
)

MedHouseVal               1.000000
MedInc                    0.688075
RoomsPerHousehold         0.209482
AveRooms                  0.151948
HouseAge                  0.105623
PopulationPerHousehold    0.065843
AveOccup                 -0.023737
Population               -0.024650
Longitude                -0.045967
AveBedrms                -0.046701
Latitude                 -0.144160
BedroomsRatio            -0.255624
Name: MedHouseVal, dtype: float64

### Insight
##### MedInc    0.688075:
Median income has the strongest positive correlation with house value (0.688).
This suggests that income level is the most important predictor of housing prices in California. Areas with higher household income tend to have significantly more expensive homes.

##### RoomsPerHousehold    0.209482:
RoomsPerHousehold shows a moderate positive correlation with house value.
Districts with more available rooms per household generally tend to have higher property values.

##### BedroomsRatio   -0.255624:
BedroomsRatio has a moderate negative correlation with house value.
Properties with a larger proportion of bedrooms relative to total rooms tend to have lower housing values.

##### PopulationPerHousehold    0.065843:
PopulationPerHousehold shows only a very weak relationship with housing prices and is unlikely to be a strong standalone predictor.

### Feature Engineering Summary

Among the engineered features, BedroomsRatio and RoomsPerHousehold provide the strongest relationships with house value.

BedroomsRatio shows a moderate negative correlation (-0.256), while RoomsPerHousehold shows a moderate positive correlation (0.209).

These results indicate that feature engineering successfully created variables with stronger predictive power than several original dataset features.